internal/nvmeof/controller/controllerserver.go

```diff
@@ -0,0 +1,563 @@
+
+	// Step 3: Delete listener
+	err = gateway.DeleteListener(ctx, nvmeofData.SubsystemNQN, nvmeofData.ListenerInfo)
+	if err != nil {
+		return fmt.Errorf("failed to delete listener for subsystem %s: %w", nvmeofData.SubsystemNQN, err)
+	}
+	log.DebugLog(ctx, "Listener deleted for subsystem %s at %s:%d", nvmeofData.SubsystemNQN,
+		nvmeofData.ListenerInfo.Address, nvmeofData.ListenerInfo.Port)
+
+	// Step 4: Delete namespace
+	if err := gateway.DeleteNamespace(ctx, nvmeofData.SubsystemNQN, nvmeofData.NamespaceID); err != nil {
+		return fmt.Errorf("failed to delete namespace %d for subsystem %s: %w",
+			nvmeofData.NamespaceID, nvmeofData.SubsystemNQN, err)
+	}
+	log.DebugLog(ctx, "Namespace %d deleted for subsystem %s", nvmeofData.NamespaceID, nvmeofData.SubsystemNQN)
+
+	// Step 5: Cleanup empty subsystem
+	if err := cleanupEmptySubsystem(ctx, gateway, nvmeofData.SubsystemNQN); err != nil {
+		return fmt.Errorf("failed to cleanup empty subsystem %s: %w", nvmeofData.SubsystemNQN, err)
+	}
+	log.DebugLog(ctx, "NVMe-oF resources cleaned up for volume with ID %s", req.GetVolumeId())
+
+	return nil
+}
+
+// populateVolumeContext adds NVMe-oF information to volume context for NodeServer.
+func populateVolumeContext(volume *csi.Volume, data *nvmeof.NVMeoFVolumeData) {
+	if volume.VolumeContext == nil {
+		volume.VolumeContext = make(map[string]string)
+	}
+	listenerPortStr := strconv.FormatUint(uint64(data.ListenerInfo.Port), 10)
+	GatewayManagementInfoPortStr := strconv.FormatUint(uint64(data.GatewayManagementInfo.Port), 10)
+
+	volume.VolumeContext["nvmeof.SubsystemNQN"] = data.SubsystemNQN
+	volume.VolumeContext["nvmeof.NamespaceID"] = strconv.FormatUint(uint64(data.NamespaceID), 10)
+	volume.VolumeContext["nvmeof.HostNQN"] = data.HostNQN
+	volume.VolumeContext["nvmeof.ListenerIpAddress"] = data.ListenerInfo.Address
+	volume.VolumeContext["nvmeof.ListenerPort"] = listenerPortStr
+	volume.VolumeContext["nvmeof.ListenerHostname"] = data.ListenerInfo.Hostname
+	volume.VolumeContext["nvmeof.GatewayAddress"] = data.GatewayManagementInfo.Address
+	volume.VolumeContext["nvmeof.GatewayPort"] = GatewayManagementInfoPortStr
+}
+
+// storeNVMeoFMetadata stores all NVMe-oF data in RBD volume metadata for cleanup operations.
+func (cs *Server) storeNVMeoFMetadata(
+	ctx context.Context,
+	req *csi.CreateVolumeRequest,
+	volumeID string,
+	nvmeofData *nvmeof.NVMeoFVolumeData,
+) error {
+	// Create RBD manager
+	mgr := rbdutil.NewManager(cs.backendServer.Driver.GetInstanceID(), nil, req.GetSecrets())
+	defer mgr.Destroy(ctx)
+
+	// Get RBD volume
+	rbdVol, err := mgr.GetVolumeByID(ctx, volumeID)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "failed to find volume with ID %q: %v", volumeID, err)
+	}
+	defer rbdVol.Destroy(ctx)
+
+	GatewayManagementInfoPortStr := strconv.FormatUint(uint64(nvmeofData.GatewayManagementInfo.Port), 10)
+	ListenerInfoPortStr := strconv.FormatUint(uint64(nvmeofData.ListenerInfo.Port), 10)
```

Comment:
```suggestion
gatewayManagementInfoPortStr := strconv.FormatUint(uint64(nvmeofData.GatewayManagementInfo.Port), 10)
listenerInfoPortStr := strconv.FormatUint(uint64(nvmeofData.ListenerInfo.Port), 10)
```

lets not use upper case to variable names